### PR TITLE
feat: handle secrets for data connectors

### PR DESCRIPTION
**Pull request by leafty**
Merging into `leafty/build-data-connectors` (feature branch).

Add support for saving and managing secrets for data connectors.

Details:
* Add API endpoints to list, update and delete saved secrets for a given data connector.

### CODE REVIEW

**Review by olevski — CHANGES_REQUESTED** [review]
Thanks for adding this Flora. It may be easier to discuss my suggestions in person rather than message back and forth.

File: components/renku_data_services/data_connectors/models.py [hunk 1]
```diff
@@ -116,3 +116,21 @@ class DataConnectorToProjectLink(UnsavedDataConnectorToProjectLink):
...
+    """Data connector secret model."""
+
+    name: str
+    user_id: str
+    data_connector_id: ULID
+    secret_id: ULID
+
+
+@dataclass(frozen=True, eq=True, kw_only=True)
+class DataConnectorSecretPatch:
```
[the remark this answers is not in the document]
Comment (olevski) [reply components/renku_data_services/data_connectors/models.py:132]: non-blocking

File: components/renku_data_services/data_connectors/db.py [hunk 2]
```diff
@@ -438,6 +443,132 @@ async def delete_link(
...
+                .where(secrets_schemas.SecretORM.id == schemas.DataConnectorSecretORM.secret_id)
+                .where(schemas.DataConnectorSecretORM.data_connector_id == data_connector_id)
+            )
+            await session.execute(stmt)
+            stmt = (
+                delete(schemas.DataConnectorSecretORM)
+                .where(schemas.DataConnectorSecretORM.user_id == user.id)
+                .where(schemas.DataConnectorSecretORM.data_connector_id == data_connector_id)
+            )
+            await session.execute(stmt)
```
[the remark this answers is not in the document]
Comment (leafty) [reply components/renku_data_services/data_connectors/db.py]: Done in 613f4dbfc3e1330b7a51cbfb880178ecc06bc09e.

File: components/renku_data_services/data_connectors/models.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (leafty) [reply components/renku_data_services/data_connectors/models.py]: Done in 8a3ea5e5e4b7ee0fbb9ec54d02afcf79263d9476.

File: components/renku_data_services/data_connectors/api.spec.yaml [hunk 3]
```diff
@@ -498,67 +495,43 @@ components:
...
+      description: Information about a credential to save for a data connector
       properties:
         name:
-          type: string
-          description: Name of the field to store credential for
-          minLength: 1
-          maxLength: 99
-        secret_id:
-          $ref: "#/components/schemas/Ulid"
+          $ref: "#/components/schemas/DataConnectorName"
```
[the remark this answers is not in the document]
Comment (leafty) [reply components/renku_data_services/data_connectors/api.spec.yaml]: Done in e1b5f979f97d6a08a334ca513036d29fd3c23559.

**Review by olevski — APPROVED** [review]
@leafty this looks good. The failing test is fixed in main so if you rebase it will pass. But this is unrelated to your code so you dont have to rebase the PR if you dont want to.

**Comment by leafty** [comment]
> @leafty this looks good. The failing test is fixed in main so if you rebase it will pass. But this is unrelated to your code so you dont have to rebase the PR if you dont want to.

Thanks, I will update the feature branch. I have kept it up-to-date with rebases during the build.

**Pull request merged**